### PR TITLE
Fixes application crashes when calculating the winning page classification category

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/7866

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

See https://github.com/brave/brave-browser/issues/7866

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 1]
```diff
@@ -872,6 +879,7 @@ std::vector<std::string> AdsImpl::GetWinningCategories() {
 
   for (const auto& page_score : page_score_history) {
     DCHECK(page_score.size() == count);
+    DCHECK(user_model_);
```
Comment (SergeyZhukovsky) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: Are you sure we don't need a check on null pointer for `user_model_`? As that is the reason of the crash. Maybe we should fail with empty `winning_categories`?

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 2]
```diff
@@ -899,6 +906,7 @@ std::vector<std::string> AdsImpl::GetWinningCategories() {
     auto it = std::find(winning_category_page_scores.begin(),
         winning_category_page_scores.end(), page_score);
     const int index = std::distance(winning_category_page_scores.begin(), it);
+    DCHECK(user_model_);
```
Comment (SergeyZhukovsky) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: the same as above. What happens in release mode if the pointer is null?

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 3]
```diff
@@ -921,6 +929,7 @@ std::vector<std::string> AdsImpl::GetWinningCategories() {
 
 std::string AdsImpl::GetWinningCategory(
     const std::vector<double>& page_score) {
+  DCHECK(user_model_);
```
Comment (SergeyZhukovsky) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: the same as above.

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 1 above)
Comment (SergeyZhukovsky) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: We've discussed this on a call, the place should never be reached when the pointer is null. Let's 
 apply the fix and reopen the issue if we still see it.

**Comment by tmancey** [comment]
CI failed on Linux, restarting CI for Linux

**Comment by tmancey** [comment]
CI failing on Linux for unrelated issues

**Comment by NejcZdovc** [comment]
checked CI with @tmancey, linux looks unrelated

**Comment by kjozwiak** [comment]
Reproduced the original issue on `macOS 10.15.3 x64` using the following build:

Brave | 1.7.26 Chromium: 80.0.3987.122 (Official Build) nightly (64-bit)
--- | ---
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | macOS Version 10.15.3 (Build 19D76)

Using the STR from https://github.com/brave/brave-core/pull/4770#issue-380967473, ran into the following crash/error:

```
[0302/112312.477864:WARNING:process_memory_mac.cc(93)] mach_vm_read(0x7ffee9edb000, 0x2000): (os/kern) invalid address (1)
[0302/112312.560187:WARNING:system_snapshot_mac.cc(42)] sysctlbyname kern.nx: No such file or directory (2)
```

* https://stats.brave.com/dashboard#crash/5e5d31254a16f2006a000abd
* https://stats.brave.com/dashboard#crash/5e5d32f1da2b9d00672c38d8

Verification **PASSED** on `macOS 10.15.3 x64` using the following build:

Brave | 1.7.32 Chromium: 80.0.3987.122 (Official Build) nightly (64-bit)
-- | --
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | macOS Version 10.15.3 (Build 19D76)

* Went through the STR mentioned via https://github.com/brave/brave-core/pull/4770#issue-380967473 and ensured that the above crash doesn't occur and the ad is being displayed.

**Pull request merged**